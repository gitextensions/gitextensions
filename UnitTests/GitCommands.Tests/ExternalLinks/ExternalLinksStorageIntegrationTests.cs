using System.Reflection;
using ApprovalTests;
using CommonTestUtils;
using FluentAssertions;
using GitCommands.ExternalLinks;
using GitCommands.Settings;
using GitUIPluginInterfaces;
using NUnit.Framework;

namespace GitCommandsTests.ExternalLinks
{
    [TestFixture]
    public class ExternalLinksStorageIntegrationTests
    {
        private ExternalLinksStorage _externalLinksStorage;

        [SetUp]
        public void Setup()
        {
            _externalLinksStorage = new ExternalLinksStorage();
        }

        [TestCase("level1_repogit_GitExtensions", 1)]
        [TestCase("level2_repodist_GitExtensions", 3)]
        [TestCase("level3_roaming_GitExtensions", 1)]
        public void Can_load_settings(string fileName, int expected)
        {
            var content = EmbeddedResourceLoader.Load(Assembly.GetExecutingAssembly(), $"{GetType().Namespace}.MockData.{fileName}.settings.xml");

            using GitModuleTestHelper testHelper = new();
            var settingsFile = testHelper.CreateRepoFile(".git", "GitExtensions.settings", content);
            using GitExtSettingsCache settingsCache = new(settingsFile);
            RepoDistSettings settings = new(null, settingsCache, SettingLevel.Unknown);

            var definitions = _externalLinksStorage.Load(settings);
            definitions.Count.Should().Be(expected);
        }

        [Test]
        public void Can_save_settings()
        {
            using GitModuleTestHelper testHelper = new();
            string settingsFile = testHelper.CreateRepoFile(".git", "GitExtensions.settings", "﻿<dictionary />");
            using GitExtSettingsCache settingsCache = new(settingsFile);
            RepoDistSettings settings = new(null, settingsCache, SettingLevel.Unknown);

            ExternalLinkDefinition definition = new()
            {
                Name = "<new>",
                Enabled = true,
                UseRemotesPattern = "upstream|origin",
                UseOnlyFirstRemote = true,
                SearchInParts = { ExternalLinkDefinition.RevisionPart.Message },
                RemoteSearchInParts = { ExternalLinkDefinition.RemotePart.URL }
            };

            _externalLinksStorage.Save(settings, new[] { definition });

            settings.Save();

            Approvals.VerifyXml(File.ReadAllText(settingsFile));
        }
    }
}
