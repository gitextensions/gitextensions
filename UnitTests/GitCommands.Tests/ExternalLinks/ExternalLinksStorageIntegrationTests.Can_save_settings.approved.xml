<dictionary>
  <item>
    <key>
      <string>RevisionLinkDefs</string>
    </key>
    <value>
      <string>&lt;?xml version="1.0" encoding="utf-16"?&gt;
&lt;ArrayOfGitExtLinkDef&gt;
  &lt;GitExtLinkDef&gt;
    &lt;Enabled&gt;true&lt;/Enabled&gt;
    &lt;LinkFormats /&gt;
    &lt;Name&gt;&amp;lt;new&amp;gt;&lt;/Name&gt;
    &lt;RemoteSearchInParts&gt;
      &lt;RemotePart&gt;URL&lt;/RemotePart&gt;
    &lt;/RemoteSearchInParts&gt;
    &lt;SearchInParts&gt;
      &lt;RevisionPart&gt;Message&lt;/RevisionPart&gt;
    &lt;/SearchInParts&gt;
    &lt;UseRemotesPattern&gt;upstream|origin&lt;/UseRemotesPattern&gt;
    &lt;UseOnlyFirstRemote&gt;true&lt;/UseOnlyFirstRemote&gt;
  &lt;/GitExtLinkDef&gt;
&lt;/ArrayOfGitExtLinkDef&gt;</string>
    </value>
  </item>
</dictionary>